### PR TITLE
add client caching by config, extract client and logging packages

### PR DESCRIPTION
**Pull request by brandonc**
## Description

Adds tfe.Client caching and extracts a client and logging package from the provider package. It doesn't make sense for each provider (sdk v2 + framework) to initialize its own client because it has to go through ping and service discovery requests. This may also slightly speed up testing since each test begins by initializing a new client.

- [x] _Update the [Change Log](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md)_
- [x] _Update the [Documentation](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md#updating-the-documentation)_

## Testing plan

Use `TF_LOG_PROVIDER=trace` to see that only one disco request is made when using the provider (terraform does another one). Previously, there were two (per provider alias!)

### CODE REVIEW

**Comment by brandonc** [comment]
Fingers crossed but this seems to have smoothed out some test times

File: internal/client/config.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	Services   *disco.Disco
+	HTTPClient *http.Client
+	TFEHost    svchost.Hostname
+	Token      string
+	Insecure   bool
+}
+
+// Key returns a string that is comparable to other ClientConfiguration values
+func (c ClientConfiguration) Key() string {
+	return fmt.Sprintf("%x %s/%v", sha256.New().Sum([]byte(c.Token)), c.TFEHost, c.Insecure)
```
Comment (nfagerlund) [inline internal/client/config.go:55]: Oh, ha, okay — I don't know if there's any way you could manage to configure an insecure and a secure client for the same hostname, but I can definitely see why we'd want them in separate slots if someone managed it.

**Review by nfagerlund — APPROVED** [review]
Okay, this is good — the layout of the code makes sense to me (i.e. having this in a new sibling package), and I gave it a try with the extra provider logging and saw exactly what you described — 0.48.0 does discovery three times (once for each muxed provider (sdkv2, framework, and low-level)), but this branch does it once.

File: internal/client/config.go (hunk 1 above)
Comment (brandonc) [reply internal/client/config.go:55]: You can add it to the provider config as `ssl_skip_verify` So you _could_ be aliasing the provider with an insecure for some odd reason.

**Pull request merged**